notes: add moderator notes

SoggySaussages opened the pull request:
Adds support for user and message context menu application commands, and adds one called 
"Notes," which displays a list of notes created for a user by staff. It also displays buttons to add,
edit, and delete notes. Modals are used to write and edit notes.

Each note includes a timestamp of the last update, and the name or user ID of the note author.

At least one moderator permission such as kick, moderate, or manage messages is required to
access and edit notes.

Maximum of 3 notes per user, and 256 characters per note.

<img width="369" alt="Screenshot 2025-04-21 at 00 33 32" src="https://github.com/user-attachments/assets/dab9dd6f-2995-4d09-9668-7160b12e5a7a" />
<img width="405" alt="Screenshot 2025-04-21 at 00 21 32" src="https://github.com/user-attachments/assets/506e4528-b17d-4765-9fe1-76854fd8f5b0" />
<img width="417" alt="Screenshot 2025-04-21 at 00 22 04" src="https://github.com/user-attachments/assets/3ee35ed6-551e-4549-bbd2-f41cd63fa2be" />
<img width="325" alt="Screenshot 2025-04-21 at 00 22 15" src="https://github.com/user-attachments/assets/00ad1061-f0cb-457b-abc6-b4c6c4a54dde" />
<img width="334" alt="Screenshot 2025-04-21 at 00 23 16" src="https://github.com/user-attachments/assets/baf980a7-87bd-48c7-9bfd-2bc7b4810bda" />
<img width="417" alt="Screenshot 2025-04-21 at 00 23 45" src="https://github.com/user-attachments/assets/23ffeac4-6708-4c2a-a738-55fd82cb24fb" />

Signed-off-by: SoggySaussages <vmdmaharaj@gmail.com>

jo3-l commented on notes/notes.go:
```diff
@@ -0,0 +1,345 @@
+	notesButtonTypeDelete
+)
+
+func (p *parsedNotes) customID(index int, buttonType notesButtonType) string {
+	args := []string{strconv.FormatInt(p.userID, 10)}
+	args = append(args, strconv.Itoa(index))
+	switch buttonType {
+	case notesButtonTypeNew:
+		args[1] = "new"
+	case notesButtonTypeEdit:
+		args = append(args, "edit")
+	case notesButtonTypeDelete:
+		args = append(args, "delete")
+	}
+
+	id := strings.Join(args, "-")
+	id = "notes_" + id
+	return id
+}
```
This code for generating the custom ID is tightly coupled to the code that decodes it (handleNoteButton and friends), but they are rather far apart in the source. What do they think about adding a function that parses data out of the custom ID right beside this function, and using it in handleNoteButton? Something like

```go
type NoteAction {
    type notesButtonType // a better name for the type might be noteActionType 
    index int // unset if action==new
}

func encodeCustomID(action noteAction, userID int64) string // replaces parsedNotes.customID
func decodeCustomID(customID string) (NoteAction, bool)    // use in handleNoteAction and friends
```

SoggySaussages replied:
Oo, I like it. Will work on this later.

jo3-l commented on notes/notes.go:
```diff
@@ -0,0 +1,345 @@
+
+const noNotesMsg = "No shared staff notes written for this user."
+
+func (p *parsedNotes) createMessageContent() *discordgo.MessageSend {
```
This is a minor nit, but I'm concerned `parsedNotes` is juggling too many concerns at the moment. It's responsible for holding parsed notes (good), for saving and modifying the corresponding database entry (good), but seemingly also for Discord-related things like generating message content, custom IDs, buttons, messages, and modals. I think those last few items can be separated into free-standing functions that accept `parsedNotes` as an argument.

Concretely, I'm wondering if

```go
type Notes struct {
    guildID, userID int
    notes []Note
}

func (*Notes) Add(...)
func (*Notes) Edit(...)
func (*Notes) Save(...)

func CreateMessageContent(*Notes)
func CreateButtons(*Notes)
// ...
```
would be a better organizational structure for this code.

SoggySaussages replied:
Yep that makes sense to me, will adjust a little later.

SoggySaussages commented:
> Nice work. I left a couple of comments below for your consideration, mainly about code and database structure as opposed to any real issues (i.e., take them with a grain of salt.)
> 
> I do have one substantial concern, having to do with the use of indexes as references to notes. Specifically, what happens in the following case?
> 
> 1. User X initially has 3 notes associated with them.
> 2. Moderator A creates a menu to manage X's notes.
> 3. Moderator B creates a _separate_ menu to manage X's notes.
> 4. Moderator B deletes note 1 from their menu.
> 5. Moderator A deletes note 1 from their (now outdated?) menu.
> 
> What happens at step 5? In particular, how many notes are deleted in total? At a glance it seems like _two_ notes will be deleted instead of just 1. That seems bad. But it's quite possible I'm missing something in the code that handles this.

Hey Joe, a good thought. Notably, whenever a note is deleted, the note in that index is simply set
to an empty note. This means if you have note one and note two, then delete note 1, now the user has Note 1 which is an empty string, and Note 2 which is untouched.

<img width="310" alt="Screenshot 2025-04-21 at 12 14 40" src="https://github.com/user-attachments/assets/36700b63-826e-4f92-906d-a03f955af4c3" />

Should two people try to delete note 1 at once, they would simply assign an empty note to index 1 twice, which shouldn't cause any issues. In a similar vein, if Moderator B deletes a note and
Moderator A tries to edit it, Moderator A would simply see a modal with an empty field and be able
to edit the note just fine. Even if moderator A already had the modal open when moderator B
deleted it, since no indexes have changed the edit would simply edit the empty note at index 1
to have text again.

Does this address the concerns you have with concurrent altering of the same user's notes?

SoggySaussages commented on notes/schema.go:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE IF NOT EXISTS notes (
+	guild_id BIGINT NOT NULL,
+	member_id BIGINT NOT NULL,
+	notes []TEXT NOT NULL,
```
This would require a big rework of the indexing system, see my response to your general 
review. I also just like only dealing with one db entry as a general rule, my days of CC coding
and minimizing database entries showing through I fear. Do you see any disadvantages to
this method that warrant the complete rework?

jo3-l replied:
It's not a particularly major concern--and I'm sure you have more experience using this system that I have--but not using indexes to refer to notes seems nicer to me, both implementation- and UX-wise. I originally wrote this response to your general comment, which is more relevant here:

> Yes, that addresses my original concern. Does that mean that, in such a case, adding a new note after the first one is deleted results in the new note using up the first slot? (It seems so from another quick glance at the code.) I'm not sure how intuitive this is.

If we didn't need to use indexes to refer to notes and instead used stable database-assigned IDs, then indexes could be implicit from 'created at' order. Then inserting a new note always goes at the end.

Of course I appreciate you've already done a lot of work with the index-based system, so this is just a suggestion. If you want to pursue my suggestion further, it might be worth getting a little more feedback from other people about how much they like the current consequences of using indexes first (in particular, add using the 'next available slot'.)

SoggySaussages replied:
I see your point, using a dynamic list instead of the
fixed array I was initially envisioning. Could a 
middle ground be sorting the notes at time of
`createMessage` before they’re displayed? This
way they still appear in chronological order without
changing the structure around too much.

ashishjh-bst replied:
why not just keep it simple and have one row per note, with all the meta data available as in who added the note. This is going to be unnecessarily complex.

ashishjh-bst replied:
Something like this. 
Otherwise there is a lot of management in code which the db could have done already. 

CREATE TABLE IF NOT EXISTS notes (
        id  BIGINT AUTO_INCREMENT, 
	guild_id BIGINT NOT NULL,
	member_id BIGINT NOT NULL,
	author_id BIGINT NOT NULL 
	created_at TIMESTAMP DEFAULT TIMESTAMP 
	note TEXT NOT NULL,
	PRIMARY KEY(guild_id, member_id)
)

ashishjh-bst replied:
@SoggySaussages  ^

SoggySaussages commented:
> Hey, awesome PR to implement the note system! But could you maybe implement that premium users get more notes?

Hey! The limit I’ve put here is just proposed, actual limits
both free and premium would need to be approved by Des.
I’ll let him give feedback on which limits he’d like to see and
then implement those. Thanks for the feedback for making 
his decision :)

egelhaus commented:
> > Hey, awesome PR to implement the note system! But could you maybe implement that premium users get more notes?
> 
> Hey! The limit I’ve put here is just proposed, actual limits both free and premium would need to be approved by Des. I’ll let him give feedback on which limits he’d like to see and then implement those. Thanks for the feedback for making his decision :)

Hey, sorry I didn't knew that Des would be making this decision, but thanks for the answer.

SoggySaussages commented:
I've rebased this to #1879 and migrated to use components v2 because it is objectively much better. I've also addressed
(or attempted to address) the remaining review notes.

![Screenshot 2025-04-24 at 16 12 58](https://github.com/user-attachments/assets/a82a95be-8184-4ee2-a644-eb9e85fd9c4a)

ennogelhaus commented:
Any idea when this will be merged?

Bennet1996 commented:
> Añade soporte para comandos de aplicación de menú contextual de usuario y mensaje, y añade uno llamado "Notas", que muestra una lista de notas creadas para un usuario por el personal. También muestra botones para añadir, editar y eliminar notas. Los modales se utilizan para escribir y editar notas.
> 
> Cada nota incluye una marca de tiempo de la última actualización y el nombre o ID de usuario del autor de la nota.
> 
> Se requiere al menos un permiso del moderador, como expulsar, moderar o gestionar mensajes, para acceder y editar notas.
> 
> Máximo de 3 notas por usuario, y 256 caracteres por nota.
> 
> <img alt="Screenshot 2025-04-21 at 00 33 32" width="369" src="https://private-user-images.githubusercontent.com/110698921/435586299-dab9dd6f-2995-4d09-9668-7160b12e5a7a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MTYyMzksIm5iZiI6MTc2NzQxNTkzOSwicGF0aCI6Ii8xMTA2OTg5MjEvNDM1NTg2Mjk5LWRhYjlkZDZmLTI5OTUtNGQwOS05NjY4LTcxNjBiMTJlNWE3YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwM1QwNDUyMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MGE1ZWU4ZmFhNmNmNTQ5ZjQwODg5ZjFlODQyODhiNjY1ZWY2ZjE0Y2ZmNmVjMDUyMmUwZTU2NWYzNzA0YmM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.GP_mAN-DPDEg1CCpSy0hfbGyKeYKWp7AUmE3V5a1vaY"> <img alt="Screenshot 2025-04-21 at 00 21 32" width="405" src="https://private-user-images.githubusercontent.com/110698921/435583870-506e4528-b17d-4765-9fe1-76854fd8f5b0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MTYyMzksIm5iZiI6MTc2NzQxNTkzOSwicGF0aCI6Ii8xMTA2OTg5MjEvNDM1NTgzODcwLTUwNmU0NTI4LWIxN2QtNDc2NS05ZmUxLTc2ODU0ZmQ4ZjViMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwM1QwNDUyMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MGZkZjUzMDFkNTY2NGQ0NzQxMDFiOGVkNzZjYTAyMGUyOGVlYzAxNzlhN2IzMzUyZGVmZjNhOTlkZWZiNjdlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.nUwSLxR_fCPw7pf9o5Ya08ac67uB3F3mSYWnSVjRiQQ"> <img alt="Screenshot 2025-04-21 at 00 22 04" width="417" src="https://private-user-images.githubusercontent.com/110698921/435583878-3ee35ed6-551e-4549-bbd2-f41cd63fa2be.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MTYyMzksIm5iZiI6MTc2NzQxNTkzOSwicGF0aCI6Ii8xMTA2OTg5MjEvNDM1NTgzODc4LTNlZTM1ZWQ2LTU1MWUtNDU0OS1iYmQyLWY0MWNkNjNmYTJiZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwM1QwNDUyMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMWU1NDAyOTI3MDkwYTU4MDM1MjI1MmUxYmFkNmYxNmYwYTdmNGZlYWJkMGJmMGE5ZGNmYTgyM2RmMjU1N2E1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.R3q5ANiF5JivfgcjqqKixN8Hk58-Y4dguY4o-BcROAA"> <img alt="Screenshot 2025-04-21 at 00 22 15" width="325" src="https://private-user-images.githubusercontent.com/110698921/435583884-00ad1061-f0cb-457b-abc6-b4c6c4a54dde.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MTYyMzksIm5iZiI6MTc2NzQxNTkzOSwicGF0aCI6Ii8xMTA2OTg5MjEvNDM1NTgzODg0LTAwYWQxMDYxLWYwY2ItNDU3Yi1hYmM2LWI0YzZjNGE1NGRkZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwM1QwNDUyMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OWE5MzQxZWI3YjZlZjAwZDYxZGY3ZGU2YmFkNDJmMTAxMjk0NDNhZDM3MmE5YWVmNzJiYWIzMjE5MGRhY2IyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bv7Gs7CNPthWBWrnRQCeqi6vFXBQ-tRpyWUN4ObDjvY"> <img alt="Screenshot 2025-04-21 at 00 23 16" width="334" src="https://private-user-images.githubusercontent.com/110698921/435583889-baf980a7-87bd-48c7-9bfd-2bc7b4810bda.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MTYyMzksIm5iZiI6MTc2NzQxNTkzOSwicGF0aCI6Ii8xMTA2OTg5MjEvNDM1NTgzODg5LWJhZjk4MGE3LTg3YmQtNDhjNy05YmZkLTJiYzdiNDgxMGJkYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwM1QwNDUyMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYWEwNzBmZDZkOGM5MTdkZWRmMjRhNmVhMDAwMmU0MjIzYjk0OWViMWIyZDA0NWI0YzRjZjVkYzU1Mjk4NjBkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.FC-JwMzF4Y6b1oVDtAyi0LCnpgsSBW73g3glb0Gqquc"> <img alt="Screenshot 2025-04-21 at 00 23 45" width="417" src="https://private-user-images.githubusercontent.com/110698921/435583896-23ffeac4-6708-4c2a-a738-55fd82cb24fb.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc0MTYyMzksIm5iZiI6MTc2NzQxNTkzOSwicGF0aCI6Ii8xMTA2OTg5MjEvNDM1NTgzODk2LTIzZmZlYWM0LTY3MDgtNGMyYS1hNzM4LTU1ZmQ4MmNiMjRmYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwM1QwNDUyMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZjBlYjZiZWQ3OWYxY2QwNjRmZTQ5NWNmMjMzMjA5ZTQyN2U3M2JmNjY5N2E0YjY0YTExZDkzYmJmMDcyYmM5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.XlD3mIAVeRn6dM4X0XusjbsAfv0k3IYDBu8_VTWqjbE">
> Despedido por: SoggySaussages [vmdmaharaj@gmail.com](mailto:vmdmaharaj@gmail.com)
